### PR TITLE
Add --no-partialeq <regex> flag

### PR DESCRIPTION
**Pull request by alexeyzab**
Related to #965.

- [x] Add a new RegexSet member to bindgen::Builder (similar to the whitelisted_types set).

- [x] A Builder method to add strings to that RegexSet.

- [x] Plumbing in src/options.rs to convert --no-partialeq <regex> CLI flags into invocations of the builder method.

- [x] Make the MonotoneFramework::constrain function in src/ir/analysis/derive_partialeq.rs check if the given item is explicitly marked not to be Partialeq, and if so, insert it into the self.cannot_derive_partialeq set via return self.insert(id).

- [x] Tests!

- [x] When the no-partialeq type is transitively referenced by a whitelisted item

- [x] When the no-partialeq type is explicitly whitelisted

- [x] When the no-partialeq type is marked opaque

This is my first pass at implementing this functionality, I haven't implemented the tests yet. I wanted to make sure I am on the right track, particularly when it comes to updating `MonotoneFramework::constrain`.

r? @fitzgen

### CODE REVIEW

File: src/ir/analysis/derive_partial_eq.rs [hunk 1]
```diff
@@ -144,6 +145,11 @@ impl<'ctx> MonotoneFramework for CannotDerivePartialEq<'ctx> {
             };
         }
 
+        let name = item.canonical_name(self.ctx);
+        if self.ctx.options().no_partialeq_types.matches(&name) {
```
Comment (fitzgen) [inline src/ir/analysis/derive_partial_eq.rs]: Let's abstract this check into a `BindgenContext::no_partialeq_by_name`, similar to `BindgenContext::is_opaque_by_name` and `BindgenContext::is_blacklisted_by_name`.

**Comment by alexeyzab** [comment]
r? @fitzgen
I've added the required tests, but I am not quite sure I did it right. I use `--with-derive-partialeq` for all three tests and then add the additional flags per your description. When I explicitly whitelist or make opaque a particular type `PartialEq` still gets derived.

**Comment by fitzgen** [comment]
It makes sense that opaque types still `derive(PartialEq)` because the `no_partialeq_by_name` check comes after the opaque check in the `constrain` function. We should move the `no_partialeq_by_name` above the opaque check.

The explicitly whitelisted types always getting `derive(PartialEq)` even when we request they shouldn't sounds more problematic. Let me look at the code a little more...

**Comment by fitzgen** [comment]
Hm yeah the test headers look how I'd expect them to, but the generated bindings don't. I'd try adding some debug `println!`s in the `CanDerivePartialEq` implementations and `lookup_item_id_can_derive_partialeq` methods. Then, I'd run with extra logging:

```
$ RUST_LOG=bindgen::ir::analysis RUST_LOG_LEVEL=trace ./tests/test-one.sh <one of your new tests>
```

If you still can't figure out what's going on after that, I can try pulling these changes down myself and investigating.

----------------------------

BTW, I'm not sure if you're aware that the ["impl period"](https://blog.rust-lang.org/2017/09/18/impl-future-for-rust.html) has just begun, but the folks hacking on `bindgen` during the impl period are hanging out in https://gitter.im/rust-impl-period/WG-dev-tools-bindgen if you want to join us :)

**Comment by alexeyzab** [comment]
> We should move the no_partialeq_by_name above the opaque check.
I've tried it just now, but the result is still the same. Makes me wonder if `no_partialeq_by_name` is doing the right thing at all. I'll try to figure it out.

Thanks for the invite! :)

**Comment by fitzgen** [comment]
@bors-servo r+

Thanks @alexeyzab !

**Pull request merged**